README.md
<div align="center" style="padding-bottom: 20px">
    <h1>Django + React + Postgres + Docker + Heroku template</h1>
    <img src="https://img.shields.io/badge/Python-14354C?style=for-the-badge&logo=python&logoColor=white" alt=""/>
    <img src="https://img.shields.io/badge/Django-092E20?style=for-the-badge&logo=django&logoColor=white" alt=""/>
    <img src="https://img.shields.io/badge/DJANGO-REST-ff1709?style=for-the-badge&logo=django&logoColor=white&color=ff1709&labelColor=gray" alt=""/>
    <img src="https://img.shields.io/badge/PostgreSQL-316192?style=for-the-badge&logo=postgresql&logoColor=white" alt=""/>
    <img src="https://img.shields.io/badge/TypeScript-007ACC?style=for-the-badge&logo=typescript&logoColor=white" alt=""/>
    <img src="https://img.shields.io/badge/React-20232A?style=for-the-badge&logo=react&logoColor=61DAFB" alt=""/>
    <img src="https://img.shields.io/badge/Docker-008FCC?style=for-the-badge&logo=docker&logoColor=white" alt=""/>
    <img src="https://img.shields.io/badge/nginx-%23009639.svg?style=for-the-badge&logo=nginx&logoColor=white" alt=""/>
    <img src="https://img.shields.io/badge/Heroku-430098?style=for-the-badge&logo=heroku&logoColor=white" alt=""/>
    <img src="https://img.shields.io/badge/github%20actions-%232671E5.svg?style=for-the-badge&logo=githubactions&logoColor=white" alt=""/>
</div>

This repository serves as a starting point for developing a
production-ready application using Django Rest Framework,
React with Typescript and Postgres in a Dockerized environment
with an option to deploy to Heroku. Running development setup
without docker-compose is also possible.

### Tools, libraries, frameworks:

This setup has been tested with Python 3.8/3.9 and Node 14.

### Backend

- `Python 3.8/9`
- `django`, `djangorestframework` Django + Django Rest Framework
- `django-cors-headers` - handling cross origin requests
- `coverage` - for code coverage reports and running unit tests
- `psycopg2` - needed to use Postgres (in Docker container)
- `gunicorn` - production wsgi http server
- `whitenoise` - building static files
- `black`, `isort`, `flake8` - for code quality (optional)

Suggested packages:

- `drf-yasg` - open api documentation (swagger and redoc)
- `django-rest-auth`, `django-allauth`, `djoser` - making auth easier
- `django-filter` - enables filtering querysets with url parameters and more
- `django-import-export` - import/export data from admin page
- `django-debug-toolbar` - useful debugging tool

### Frontend

- `Node 14`
- `react` - React
- `typescript` - Typescript
- `react-router-dom` - frontend routing
- `axios` - for making HTTP requests

Suggested packages:

- UI libraries such as `Chakra-UI`, `Material-UI`, `Reactstrap`, `TailwindCSS` etc.
- `react-query` - very useful package which handles data fetching logic for you
- `formik` + `yup` - handling form state and validation
- `@reduxjs/toolkit` + required packages
  (react-redux, redux etc.) - library which makes Redux much easier to understand and use
- `cypress` - for e2e testing

# Development setup

## Without Docker

### Backend

Create a virtual environment from cmd (or do it in Pycharm manually)

```shell script
cd backend
python -m venv venv
```

> Note: please adjust the command with the `python` executable on your
> computer, because sometimes (example: on Ubuntu or macOS) Python 3
> can only be executed using `python3`, not `python`.

Activate the virtual environment that was just created
On Windows:

```shell
venv\Scripts\activate
```

On Linux/macOS:

```shell
source venv/bin/activate
```

If successful, there should be `(venv)` in your cmd/terminal prompt.

Install the required packages with the following command.

```shell script
python -m pip install --upgrade pip
pip install -r requirements.txt
```

Run django application from cmd (or add new Django configuration if using Pycharm)

```shell script
python manage.py runserver
```

Preparing (if there are any changes to db schema) and running migrations

```shell script
python manage.py makemigrations
python manage.py migrate
```

Create superuser

```shell script
python manage.py createsuper user
```

### Frontend

Install node dependencies.

```shell script
cd frontend
npm install
```

Run development server in second terminal

```shell script
npm start
```

### Backend tests coverage

```shell script
cd backend
```

Run tests using Coverage (unit tests + report) instead of `python manage.py test`

```shell script
coverage run manage.py test
```

Get report from coverage:

```shell script
coverage report -m
```

## With Docker

First define environmental variables in `.env` in root directory:

```
DB_NAME
DB_USERNAME
DB_PASSWORD
```

Make sure Docker Engine is running.

While in **root directory**, build docker images and run them with docker-compose. This might take up to few minutes.
Rebuilding image is crucial after installing new packages via pip or npm.

```shell script
docker-compose -f docker-compose.dev.yml up --build
```

Application should be up and running: backend `127.0.0.1:8000`, frontend `127.0.0.1:3000`, and nginx `127.0.0.1:8080`.

If images had been installed and **no additional packages have been installed**, just run to start containers:

```shell script
docker-compose -f docker-compose.dev.yml up
```

Bringing down containers with **optional** -v flag removes **all** attached volumes and invalidates caches.

```shell script
docker-compose -f docker-compose.dev.yml down
```

To run commands in active container:

```shell script
docker exec -it <CONTAINER_ID/CONTAINER_NAME> <command>
```

In case there is no bash installed, try:

```shell script
docker exec -it <CONTAINER_ID/CONTAINER_NAME> /bin/sh
```

e.g

```shell script
docker exec -it backend python manage.py createsuperuser
docker exec -it backend coverage run manage.py test
docker exec -it frontend /bin/sh
```

## CI/CD

This repository uses Github Actions to run test and deployment pipeline.  
`tnd.yml` - runs backend-frontend test separately.

There's also have automatic deploys application to Heroku. You need the autorization token for the deployment.

```shell
heroku create {your-app-name}
```

Make sure you are logged in to Heroku.
Generate an API token for your Heroku profile

```shell
heroku authorizations:create
```

Furthermore, There's success/failure notification _discord_ action that make teammates or another developer's work easier to handles mis-communication about merging PRs, as projects become more complex, so too does the process of building and testing the code.

`code-quality.yml` - runs backend and frontend code quality separately.

# Production Deployment

1.  [Create Heroku Account](https://signup.heroku.com/dc)
2.  [Download/Install/Setup Heroku CLI](https://devcenter.heroku.com/articles/heroku-cli#download-and-install)

    - After install, log into Heroku CLI: `heroku login`

3.  Run: `heroku create <app name>` to create the Heroku application
4.  Set your environment variables for your production environment by running:

    ```
    heroku config:set VARIABLE=value
    ```

    Variables to set: `DJANGO_SETTINGS_MODULE=core.settings.prod`,
    `DJANGO_SUPERUSER_USERNAME`, `DJANGO_SUPERUSER_PASSWORD`, `PRODUCTION_HOST=<app name>.herokuapp.com`,
    `SECRET_KEY`,

5.  Run: `heroku stack:set container` so Heroku knows this is a containerized application
6.  Run: `heroku addons:create heroku-postgresql:hobby-dev` which creates the postgres add-on for Heroku
7.  Deploy app by running: `git push heroku master/main`,  
    _or_ manually in Heroku dashboard  
    _or_ by pushing to your github repository, having Automatic Deploys set up
8.  Go to `<app name>.herokuapp.com` to see the published website.

### Run commands in Heroku:

```shell
heroku run bash --app <your_app_name>
```

e.g

```shell
heroku run bash --app django-react-heroku-test
~ $ python backend/manage.py createsuperuser
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.2](https://github.com/Recedivies/django-react-template/compare/v0.1.1...v0.1.2) (2022-07-12)


### Bug Fixes

* tnd pipeline and dev settings ([1d71053](https://github.com/Recedivies/django-react-template/commit/1d71053e9da2afa82203d8a87efa616af4f447d6))

### [0.1.1](https://github.com/Recedivies/django-react-template/compare/v0.1.0...v0.1.1) (2022-07-08)


### Bug Fixes

* commits log ([0988e3b](https://github.com/Recedivies/django-react-template/commit/0988e3bd76ece41bd808bbb308642070ab326528))

## [0.1.0](https://github.com/Recedivies/django-react-template/compare/v0.0.4...v0.1.0) (2022-07-08)


### Features

* added a healh check simple endpoint ([a52c744](https://github.com/Recedivies/django-react-template/commit/a52c744f9411684bd9e854206c1ef1de9d58aa4d))


### Bug Fixes

* code format ([8c815ef](https://github.com/Recedivies/django-react-template/commit/8c815ef587c7692f5b01639aac55d7de49f098ea))
* rollback false ([c30e1b9](https://github.com/Recedivies/django-react-template/commit/c30e1b9b4243a6dcd1987d4b889b18f2780b6615))
* test fe ([fae5e95](https://github.com/Recedivies/django-react-template/commit/fae5e9509cced6deb206e4cdf61b46bff3d6124a))
* trigger auto deploys ([d6b3c67](https://github.com/Recedivies/django-react-template/commit/d6b3c67d6665a4d4c7ddd0cc2938c8fa2af8e674))

### [0.0.4](https://github.com/Recedivies/django-react-template/compare/v0.0.3...v0.0.4) (2022-07-08)


### Features

* added ci/cd pipeline ([59e58d8](https://github.com/Recedivies/django-react-template/commit/59e58d8bcc9c091f6f1b5a5e5e59550cf0f65426))

### [0.0.3](https://github.com/Recedivies/django-react-template/compare/v0.0.2...v0.0.3) (2022-07-08)


### Features

* added docker configuration ([cee96b4](https://github.com/Recedivies/django-react-template/commit/cee96b4e968cc7c6cfc280edecb4d60d979ce0b7))
* added heroku configuration ([31e9ef3](https://github.com/Recedivies/django-react-template/commit/31e9ef3a3bcf570d8b1a1ccf6e3d90be6027a77d))
* added nginx configuration ([c3ebc67](https://github.com/Recedivies/django-react-template/commit/c3ebc67a469df8b7cbbd4c1af659dfe18c7502a2))
* added simple api ([f11b1db](https://github.com/Recedivies/django-react-template/commit/f11b1db13cba4eeb55b72d1c442a54869844caee))
* added wait_for_db command ([644a65a](https://github.com/Recedivies/django-react-template/commit/644a65ab332639572310a553b9e57c6fba9726ed))
* addede code quality fe ([2e1552d](https://github.com/Recedivies/django-react-template/commit/2e1552da616c9b1652164c2b3da1d5ed940b3dd1))
* setup code quality ([379799a](https://github.com/Recedivies/django-react-template/commit/379799a285b000ef6819e3b0058f01138d62facf))
* setup django settings ([fa83790](https://github.com/Recedivies/django-react-template/commit/fa83790282eb7282b22ade25f477350abd687e31))
* test connection api ([9804ca9](https://github.com/Recedivies/django-react-template/commit/9804ca9cf25ecbfb7a893bde6bb32241004396eb))


### Bug Fixes

* axios baseUrl to connect api ([898f62c](https://github.com/Recedivies/django-react-template/commit/898f62c89115a467ea413dff23beb475653cae8d))
* base Url fe and proxy ([56b1cde](https://github.com/Recedivies/django-react-template/commit/56b1cdeaa340ba3b17d19764d6f34f38f215931c))

### [0.0.2](https://github.com/Recedivies/django-react-template/compare/v0.0.1...v0.0.2) (2022-07-07)

### 0.0.1 (2022-07-07)


### Features

* init project ([ec0ed4a](https://github.com/Recedivies/django-react-template/commit/ec0ed4a0fe9af1d4817b1ed2aedae671bcd80182))


backend/core/settings/dev.py
"""
Configuration for development with Docker.
"""
import os

from core.settings.base import *  # noqa: F403, F401

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.environ.get("SECRET_KEY", "development")

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = ["backend", "localhost", "127.0.0.1"]

CORS_ALLOWED_ORIGINS = [
    "http://localhost:3000",
]

CSRF_TRUSTED_ORIGINS = [
    "http://localhost:8080",
    "http://127.0.0.1:8080",
]

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": os.environ.get("DB_NAME", "postgres"),
        "USER": os.environ.get("DB_USER", "postgres"),
        "PASSWORD": os.environ.get("DB_PASSWORD", "postgres"),
        "HOST": os.environ.get("DB_HOST", "db"),
        "PORT": os.environ.get("DB_PORT", 5432),
    }
}
